### ISSUE TITLE
org.jboss.cdi.tck.tests.deployment.discovery.EmptyBeansXmlDiscoveryTest breaks backward compat

### ISSUE THREAD

**Issue by struberg** [jakartaee/cdi-tck#407]
Can someone please explain to me the rational behind ```org.jboss.cdi.tck.tests.deployment.discovery.EmptyBeansXmlDiscoveryTest```?

An empty beans.xml file always behaved as marker file for a ALL!

What you wrote in CDI-4.0 is that "A new beans.xml 4.0 schema file has been added and the namespace of the beans_4_0.xsd schema file is xmlns:jakartaee="https://jakarta.ee/xml/ns/jakartaee", the same as 3.0. The key changes in the schema are to make the bean-discovery-mode attribute default to annotated and to use annotated as the default when an empty beans.xml is seen in a deployment. "

So if a beans.xml with a ```version="4.0"``` gets detected then the bean discovery mode gets changed. Otherwise not. 

If you really intend to change the default in CDI-4.0 for a completely empty beans.xml file then it will not even be backward compat with CDI-3.0. That would imo be a major pain for real world adoption and a very stupid move.

**Comment by manovotn** [comment]
The test is changed to correspond to changes summarized in https://github.com/jakartaee/cdi/issues/500

The initial description as well as my comment (https://github.com/jakartaee/cdi/issues/500#issuecomment-1405221674) should explain the current state in CDI 4.
In short, yes this is breaking for fraction of CDI applications and this was expected.
The above is also a reason why there is also a mandatory compatibility switch for EE integrators. However, the TCK is unable to test this as the config is implementation specific.

Since the test IMO reflects current specification requirements, I'll close this issue.
If you mean to challenge the design as such, I suggest we keep the conversation in one place - the [CDI issue](https://github.com/jakartaee/cdi/issues/500).

**Comment by struberg** [comment]
This breaks backward compatibility. Got the requirement for backward compatibility with deprecation and pruning phases ditched in JakartaEE. If so I will escalate this to the council. It might not be a failure of the TCK, but from the CDI spec though

**Comment by struberg** [comment]
BTW, the Jakarta rules explicitly forbid such changes.
https://eclipse-ee4j.github.io/jakartaee-platform/CompatibilityRequirements

**Comment by dblevins** [comment]
I do agree with @struberg that the backwards incompatibility is unfortunate, hence we did not vote +1 on CDI.  That said it is now part of the ratified spec.

The TCK test challenge process explicitly calls out that the spec itself cannot be challenged.   See "Invalid Challenges", bullet two:

 - https://jakarta.ee/committees/specification/tckprocess/

Can we agree that what's being challenged is the spec itself?  If so, we can keep this issue open but we'd need to move it to the spec repo and address it through the spec process in a future version.  I would definitely recommend a flag to portably turn on the CDI 3.0 and before behavior.

**Comment by manovotn** [comment]
> Can we agree that what's being challenged is the spec itself? If so, we can keep this issue open but we'd need to move it to the spec repo and address it through the spec process in a future version. I would definitely recommend a flag to portably turn on the CDI 3.0 and before behavior.

@dblevins this flag is anchored in the spec text of 4.0 and required to be implemented by EE containers.

To quote, [the spec says](https://jakarta.ee/specifications/cdi/4.0/jakarta-cdi-spec-4.0.html#bean_archive_full):

> For compatibility with CDI versions prior to 4.0, CDI Full products must contain an option that causes an archive with empty beans.xml to be considered an explicit bean archive.

**Comment by dblevins** [comment]
@manovotn Understood.  I referred to a portable flag vs an implementation-specific flag.  We could then test it, users can also get the benefit.  If I can think of any good options and find the time to contribute language/tests, I'll create a new ticket.

**Comment by manovotn** [comment]
True, this has come up during development and I agree a "portable flag" would be a better option.
However, CDI itself doesn't have an app wide config mechanism through which we could introduce such a flag.
We'd have to introduce some such mechanism (since I assume we don't want to bring in dep to something like MP config) for just this one option.

How we landed on this was partly because of what I said above and partly because previous versions of the spec had *exactly this kind of flag* for years. Since CDI 1.1, there was [this section in the spec](https://jakarta.ee/specifications/cdi/2.0/cdi-spec-2.0.html#bean_archive) which stated:

> For compatibility with Contexts and Dependency 1.0, products must contain an option to cause an archive to be ignored by the container when no beans.xml is present.

Which also couldn't be tested and was vendor specific but was mandated by the spec nonetheless.

But maybe we could introduce a way to assert the new flag. Perhaps adding a test that asserts the old behavior and a TCK API/SPI that will have to be implemented by whoever runs the TCK? We already have some such SPIs in TCK, the [`Contexts`](https://github.com/jakartaee/cdi-tck/blob/master/api/src/main/java/org/jboss/cdi/tck/spi/Contexts.java) one being an example.
